### DISCUSSION TITLE
Discussion about licenses

### DISCUSSION THREAD

**Discussion by MKdays** [pyinstaller/pyinstaller#5499] [PyInstaller]
I hope there are as many opinions as possible in this discussion.
Because learners focus only on implementing the function and there is little discussion about licensing.

Suppose you did not use the onefile mode.
And not all of us are lawyers.

If you freeze your module with a Pyinstaller, it will probably work as follows.

1) Convert **_Python/Lib_** to _**base_library.zip**_ and bundle it.
2) files (**_*.pyd and *.dll_**) are bundled from **_Python/DLLs_**
3) Python Core files (_**python3.dll**_ and **_python39.dll_**) are bundled.
4) MS C RUNTIME LIBRARY (**_MSVCP140.dll, VCRUNTIME140.dll_**) are bundled.
5) The **_*.py files_** on the module being imported are directly dependent on exe.

## The obligation of licensing notices to use Python libraries
Add the following statement to the program to notify you of the PSF-2.0 license and display the copyright holder.

> Python : https://www.python.org/
> Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006 Python Software Foundation; All Rights Reserved
> Licensed under the PSF-2.0

Should attach a copy of the license. The document is located in Python/LICENSE.txt

## MS C RUNTIME LIBRARY
I received the answer for this part as below. I'm not sure about the answer either. Please refer to the link below.

Please refer to the below.
https://answers.microsoft.com/en-us/windows/forum/windows_other-windows_programs/can-i-redistribute-the-dll-for-the-window/754f1a8f-a07e-4fff-954f-74b2ceb2503d?tm=1610874789425

## Compliance with LGPL
Perhaps the most controversial part.

My opinion is that if Python source code is compiled into exe binary, LGPL cannot be complied with. The reason is as follows.

First,
Python imports external modules without linking them.
Conceptually, it can be thought of as linking, but if it is a legal interpretation, the story may be different.
Consequently, compliance with the use of LGPL as a library itself becomes ambiguous.

Second,
LGPL requires that the distributed user can replace the library.
As described above, *.py files are bundled into exe. The dll is replaceable, but the source code included in exe cannot be modified by the distribution recipient.

Please refer to the below.
https://stackoverflow.com/questions/8580223/using-python-module-on-lgpl-license-in-commercial-product
https://velovix.github.io/post/lgpl-gpl-license-compliance-with-pyinstaller/
https://github.com/pyinstaller/pyinstaller/issues/4471
https://www.qt.io/faq/tag/qt-open-source-licensing
https://kivy.org/doc/stable/guide/licensing.html

I don't think I'm 100% right.
Could you please tell me if I'm mistaken or not?

My only interest is to comply with LGPL while maintaining an exclusive license for my program.

**Comment by bwoodsend** [comment]
Reading the Python license I think it's pretty clear: That Python copyright notice needs to be visible somewhere on anything derived from Python.

The MSVC DLLs have to distributable otherwise Visual Studio is nothing but a giant legal bomb factory. Anything built by it is linked against them and, as Windows comes with almost no base libraries itself, they have to be bundled.

LGPL is a nightmare of a license. It requires you to be able to slice out and drop in replacements to shared libraries. Now I can't think of any shared library anywhere on anything that can be replaced with another library that isn't just a different version of itself without breaking the parent program. This leaves only one option: to make your program buildable by any user from source so they can eliminate dependencies which by extension mandates that it be open-source and thereby contradicts the license's own purpose - to allow commercial applications. So as far as I see it, LGPL is just a convoluted GPL. What am I missing here?

The next question is where do we put these licenses? It'd be invisible if embedded in the executable itself so I think it would have to be a separate file which the user is required to ship alongside their application. Unless they either add a *show licenses* CLI option or GUI button in which case they should be able to opt for the licenses to be packed as an accessible data resource.

**Comment by maldag** [comment]
Considering the LGPL part it should be possible to change the dll *of that exact version* to something else and link it to the application. Would that work with PyInstaller? 
This refers to my personal information the LGPL licensed library has to be *exactly the same* with the desired changes. 

Putting this into a practical idea: Can pyInstaller bundle everything into one file (using *onefile* argument), but link the respective DLLs at runtime?

**Reply by rokm** [reply]
> Would that work with PyInstaller?

It would be straightforward for `onedir` since DLLs are always on the disk and you can replace them as you want (well, except on macOS, where its dependent library paths may need to be adjusted, if those are also part of frozen bundle). 

Not so much for `onefile`, because the executable's embedded CArchive would need to be unpacked, the DLL swapped, and then repacked (following the steps that PyInstaller's executable assembly pipeline uses). It's rather tedious to do manually, but I imagine one could write a script that does this and swaps the desired components from the embedded archive. 

Of course, your python code itself could also implement provisions for allowing DLL overrides from external directories. E.g., by modifying environment (PATH on Windows, LD_LIBRARY_PATH on linux) to include an external DLL directory, or calling `AddDllDirectory` via `ctypes` on Windows. Of course, this would need to be done at very beginning, before python packages (or their extensions) that make use of those DLLs are loaded.

Replacing whole LGPL'd python packages is probably even more hassle, because you need to repack PYZ archive from executable (for both `onedir` and `onefile`) from the embedded CArchive. So it might be easier to implement option to extend `sys.path`  in your python code to allow loading of external python packages. This approach is described in one of posts that the OP linked.

Or, depending on the application itself, it might also be possible to extract the application's .pyc files (and other resources that do not come from 3rd party packages) from the executable, and then run either them in a custom/modified python environment, or rebuild the whole thing into an executable again, using that custom environment.

**Comment by dineshr93** [comment]
Hi @MKdays , Please make sure if the LGPL part of the source gets into the exe. It may be the case that LGPL part of source might not be packed inside the exe file, In that case we dont need not worry about the LGPL license.

**Reply by darthwalsh** [reply]
You can validate for yourself: the python files that will be bundled into the EXE by running `pyi-archive_viewer --log --recursive --brief PYZ-00.pyz`-- it will show a lot of python3 core libraries, and any third party dependencies that are packed into the EXE.

**Comment by darthwalsh** [comment]
It's very reasonable to be worried about this problem with LGPL \*.py files meaning you need to include the source instructions to replace those \*.py files.

Luckily, there's a relatively straightforward technical workaround for moving only the \*.py scripts out of the EXE, and bringing it along as data in a separate folder. It's analogous to choosing to "dynamically" link the PY script file, instead of the pyinstaller default of "statically" linking by bundling.

1. If you aren't already [using pyinstaller spec files](https://pyinstaller.org/en/stable/spec-files.html#spec-file-operation), switch to that.
2. The generated SPEC file should have a line of python code like `pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)`
3. `a.pure` contains every python file that would be bundled. We need to filter out the LGPL files!
    i. i.e. it will contain `p3env/lib/python3.8/site-packages/PySide6/__init__.py` which seems to be LGPL.
5. Use this code instead. This prevents pyside or shiboken files from being bundled, and instead copies them as data.

```python
to_dynamic_link = {"PySide6", "shiboken6"}
to_include = []
for bundle in a.pure:
    name, path, _ = bundle
    if name in to_dynamic_link:
        py_file = path.split("site-packages")[-1][1:]
        print("Dynamic linking", name, path, "->", py_file)
        a.datas.append((py_file, path, "DATA"))
    else:
        to_include.append(bundle)

pyz = PYZ(to_include, a.zipped_data, cipher=block_cipher)
```

Of course, your mileage will vary depending on which library you are using. In this case, the SPEC copies `PySide6/__init__.py` into the right folder, so at app runtime `from PySide6 import QtGui` will run the required logic from `__init__.py`
